File: backend/domain/documentmodel/model_test.go
package documentmodel

import (
	"errors"
	"testing"
)

func validDocument() *Document {
	return &Document{SchemaVersion: SchemaVersion, Settings: Settings{PageSize: "A4", Orientation: "portrait"}, Root: Root{Pages: []Page{{ID: "page-1", Width: A4WidthDU, Height: A4HeightDU, Children: []Node{{ID: "title", Kind: "text", Role: "element", Geometry: Geometry{Width: 1000, Height: 500}, LayoutMode: "fixed", Visibility: "shown"}}, ChildIDs: []string{"title"}}}}}
}

func TestValidateV5Document(t *testing.T) {
	if err := Validate(validDocument()); err != nil {
		t.Fatalf("valid document rejected: %v", err)
	}
}

func TestParseRejectsUnknownFieldsAndWrongVersion(t *testing.T) {
	_, err := Parse([]byte(`{"schema_version":5,"root":{"pages":[],"unexpected":true},"settings":{"page_size":"A4","orientation":"portrait"}}`))
	if err == nil {
		t.Fatal("expected unknown field error")
	}
	_, err = Parse([]byte(`{"schema_version":4}`))
	if !errors.Is(err, ErrSchemaVersion) {
		t.Fatalf("expected schema error, got %v", err)
	}
}

func TestValidateFlowFrameRequiresStoryAndOrderedChildren(t *testing.T) {
	d := validDocument()
	d.Root.Pages[0].Children[0].Role = "flow-frame"
	d.Root.Pages[0].Children[0].Kind = "flow-frame"
	d.Root.Pages[0].Children[0].LayoutMode = "flow-frame"
	if err := Validate(d); err == nil {
		t.Fatal("expected missing story error")
	}
	d.Stories = []Story{{ID: "story-1", Kind: "rich-text"}}
	d.Root.Pages[0].Children[0].StoryID = "story-1"
	d.Root.Pages[0].ChildIDs = []string{"other"}
	if err := Validate(d); err == nil {
		t.Fatal("expected child order error")
	}
}

func TestValidateRejectsUnknownMasterReference(t *testing.T) {
	d := validDocument()
	d.Root.Pages[0].MasterID = "missing"
	if err := Validate(d); err == nil {
		t.Fatal("expected missing master error")
	}
}

func TestValidateRejectsUnknownWidgetKind(t *testing.T) {
	doc := &Document{SchemaVersion: SchemaVersion, Settings: Settings{PageSize: "A4", Orientation: "portrait"}, Root: Root{Pages: []Page{{ID: "p", Width: A4WidthDU, Height: A4HeightDU, ChildIDs: []string{"n"}, Children: []Node{
		{ID: "n", Kind: "iframe", Role: "element", Geometry: Geometry{Width: 1000, Height: 1000}, Visibility: "shown", LayoutMode: "fixed"},
	}}}}}
	if err := Validate(doc); err == nil {
		t.Fatal("expected unknown widget kind to be rejected")
	}
}

File: frontend/src/builder/v5/model.test.ts
import { describe, expect, it } from 'vitest'
import {
  A4_HEIGHT_DU,
  A4_WIDTH_DU,
  du,
  validateV5,
  V5_SCHEMA_VERSION,
  type V5Document,
} from './model'
import { parseV5, serializeV5 } from './serialization'
import { largeTableFixture } from './fixtures'

const fixture = (): V5Document => ({
  schema_version: V5_SCHEMA_VERSION,
  settings: { page_size: 'A4', orientation: 'portrait' },
  root: {
    pages: [
      {
        id: 'p1',
        width: du(A4_WIDTH_DU),
        height: du(A4_HEIGHT_DU),
        margin: { top: du(0), right: du(0), bottom: du(0), left: du(0) },
        child_ids: [],
        children: [],
      },
    ],
  },
})

describe('V5 model', () => {
  it('validates and round-trips a page fixture', () =>
    expect(parseV5(serializeV5(fixture())).schema_version).toBe(5))
  it('rejects non-A4 geometry', () =>
    expect(
      validateV5({ ...fixture(), root: { pages: [{ ...fixture().root.pages[0], width: du(1) }] } }),
    ).toContain('dimensions'))
  it('provides the 500-row baseline fixture without editor state', () => {
    const story = largeTableFixture().stories?.[0]
    expect(story).toBeDefined()
    expect((story!.content as { rows: unknown[] }).rows).toHaveLength(500)
  })
})

describe('V5 flow-frame continuation validation', () => {
  const frameDoc = (overrides: Record<string, unknown>): V5Document => {
    const base = fixture()
    base.stories = [{ id: 'story-1', kind: 'rich-text', content: { text: 'hi' } }]
    base.root.pages[0].child_ids = ['frame']
    base.root.pages[0].children = [
      {
        id: 'frame',
        kind: 'flow-frame',
        role: 'flow-frame',
        story_id: 'story-1',
        geometry: { x: du(0), y: du(0), width: du(100), height: du(100), rotation: 0 },
        layout_mode: 'flow-frame',
        locked: false,
        visibility: 'shown',
        optional: false,
      },
    ]
    Object.assign(base.root.pages[0].children[0], overrides)
    return base
  }
  it('rejects an unknown continuation master', () => {
    const doc = frameDoc({ continuation: 'auto-pages', continuation_master_id: 'missing' })
    expect(validateV5(doc)).toMatch(/missing continuation master/)
  })
  it('rejects an invalid continuation policy', () => {
    const doc = frameDoc({ continuation: 'magic' })
    expect(validateV5(doc)).toMatch(/invalid continuation policy/)
  })
  it('accepts a valid continuation master reference', () => {
    const doc = frameDoc({ continuation: 'auto-pages', continuation_master_id: 'master-1' })
    doc.root.masters = [{ id: 'master-1', child_ids: [], children: [] }]
    expect(validateV5(doc)).toBeNull()
  })
})
